xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -300,15 +412,92 @@ public void run() {
     }
   }
 
-  // https://github.com/google/error-prone/issues/1767
-  @SuppressWarnings("ModifyCollectionInEnhancedForLoop")
-  private class ConfigWatcherImpl implements ConfigWatcher {
+  private class ResolveState implements LdsResourceWatcher {
+    private final ConfigOrError emptyServiceConfig =
+        serviceConfigParser.parseServiceConfig(Collections.<String, Object>emptyMap());
+    private final ResolutionResult emptyResult =
+        ResolutionResult.newBuilder()
+            .setServiceConfig(emptyServiceConfig)
+            // let channel take action for no config selector
+            .build();
+    private boolean started;
     private Set<String> existingClusters;
+    @Nullable
+    private String rdsResource;
+    @Nullable
+    private RdsResourceWatcher rdsWatcher;
+
+    @Override
+    public void onChanged(LdsUpdate update) {
+      if (!started) {
+        return;
+      }
+      logger.log(XdsLogLevel.INFO, "Receive LDS resource update: {0}", update);
+      List<VirtualHost> virtualHosts = update.getVirtualHosts();
+      String rdsName = update.getRdsName();
+      if (rdsName != null && rdsName.equals(rdsResource)) {
+        return;
+      }
+      cleanUpRdsWatcher();
+      if (virtualHosts != null) {
+        updateRoutes(virtualHosts);
+      } else {
+        rdsResource = rdsName;
+        rdsWatcher = new RdsResourceWatcherImpl();
+        logger.log(XdsLogLevel.INFO, "Start watching RDS resource {0}", rdsResource);
+        xdsClient.watchRdsResource(rdsResource, rdsWatcher);
+      }
+    }
 
     @Override
-    public void onConfigChanged(ConfigUpdate update) {
+    public void onError(Status error) {
+      if (!started) {
+        return;
+      }
+      logger.log(
+          XdsLogLevel.WARNING,
+          "Received error from xDS client {0}: {1}", xdsClient, error.getDescription());
+      listener.onError(error);
+    }
+
+    @Override
+    public void onResourceDoesNotExist(String resourceName) {
+      if (!started) {
+        return;
+      }
+      logger.log(XdsLogLevel.INFO, "LDS resource {0} unavailable", resourceName);
+      cleanUpRdsWatcher();
+      listener.onResult(emptyResult);
+    }
+
+    private void start() {
+      if (started) {
```

Comment:
I believe `start()` is only called right after instantiation and won't be called twice.

Reply:
The problem for not using `start()`/`stop()` is that the `LdsResourceWatcher` implementation class will not be self-contained. It would be something like

```java
class XdsNameResolver {
  private LdsResourceWatcher ldsWatcher;

  @Override
  public void start(Listener2 listener) {
    ...
    ldsWatcher = new LdsResourceWatcherImpl();
    xdsClient.watchLdsResource(authority, ldsWatcher);
  }


  @Override
  public void shutdown() {
    ...
    if (xdsClient != null) {
      if (ldsWatcher != null) {
        xdsClient.cancelLdsResourceWatch(authority, ldsWatcher);
      }
      ...
    }
    ...
  }
}

```

I prefer the existing approach as it encapsulate things better. What do you think?

Reply:
I was not saying not to use `start()/stop()`. It was about checking the `started` flag.